Refactor: Extract SRP input from create vault component

Gudahtt opened the pull request:
This is a pure refactor that extracts the SRP input from the `CreateNewVault` component. This is intended to make future changes to the SRP input easier, and to reduce duplication between the old and new onboarding flows.

Extensive unit tests have been added for the new SRP input component.

A new test library was added (`@testing-library/user-event`) for simulating user events with components rendered using the `@testing-library` library.

A new helper method has been added (`renderWithLocalization`) for rendering components using `@testing-library` with just our localization contexts added as a wrapper. The localization contexts were already added by the `renderWithProviders` helper function, but there is no need for a Redux provider in these unit tests.

Manual testing steps:  
  - Test all functionality of the "import" page of onboarding
  - Test all functionality of the "Restore vault" page that is reachable from the lock screen

Gudahtt committed:
Refactor: Extract SRP input from create vault component

This is a pure refactor that extracts the SRP input from the
`CreateNewVault` component. This is intended to make future changes to
the SRP input easier, and to reduce duplication between the old and new
onboarding flows.

Extensive unit tests have been added for the new SRP input component.

A new test library was added (`@testing-library/user-event`) for
simulating user events with components rendered using the
`@testing-library` library.

A new helper method has been added (`renderWithLocalization`) for
rendering components using `@testing-library` with just our
localization contexts added as a wrapper. The localization contexts
were already added by the `renderWithProviders` helper function, but
there is no need for a Redux provider in these unit tests.

danjm commented on ui/components/app/srp-input/srp-input.js:
```diff
@@ -0,0 +1,112 @@
+import { ethers } from 'ethers';
```
This looks to be a clean refactor and extraction of the code that was in ui/components/app/create-new-vault/create-new-vault.js

Gudahtt commented on ui/components/app/create-new-vault/create-new-vault.js:
```diff
@@ -140,56 +108,7 @@ export default function CreateNewVault({
-            <Typography tag="span">{t('showSeedPhrase')}</Typography>
-          </label>
-        </div>
+        <SrpInput seedPhrase={seedPhrase} onChange={setSeedPhrase} />
```
Oops! You're right, the SrpInput maintains its own internal draft and just communicates that outward. I can remove this. I don't know of a reason to set the draft SRP from outside the component, and to do that while avoiding thrashing as it changed might be complex, so I will not attempt to do that.

```suggestion
        <SrpInput onChange={setSeedPhrase} />
```

Gudahtt committed:
Remove unused prop

Co-authored-by: Dan J Miller <danjm.com@gmail.com>

Gudahtt commented on ui/components/app/srp-input/srp-input.test.js:
```diff
@@ -0,0 +1,539 @@
+  `${new Array(11).fill('test').join('  ')}  ball`,
+  `${new Array(11).fill('test').join('\t')}\tball`,
+];
+
```
>because it has the incorrect checksum

This is what the `invalidCorrectLength` test is for. That's an SRP that is invalid because it fails the checksum, but it is the correct length.

Gudahtt replied:
We could add a test case with a word not on the wordlist :thinking: That's a good idea. I was also considering adding tests for the other locales we support for the SRP.

danjm replied:
Oh, I guess that is what invalidCorrectLength does

Gudahtt replied:
Just added a test that includes a SRP with an invalid word, but a correct checksum.

I generated this word using this project: https://github.com/avsync/bip39chk
I altered the wordlist to include "aardvark", then I used that script to find a valid checksum word for the SRP with aardvark in it.

I tried adding a test for non-English locales as well, but apparently we only support English right now. The library we use to generate private key from the mnemonic in the background supports other locales, but the validation we're doing in the UI only works for English. So we can revisit that later if/when we expand it to support other locales.

danjm commented on ui/components/app/srp-input/srp-input.test.js:
```diff
@@ -0,0 +1,539 @@
+  `${new Array(11).fill('test').join('\t')}\tball`,
+];
+
+describe('srp-input', () => {
```
All these tests look really good :+1:

danjm reviewed (dismissed):
This looks good to me. There is a suggestion to add one more test case, but that doesn't need to be a blocker.

Gudahtt committed:
Add test with invalid word

A test has been added to ensure a SRP with a correct checksum but an
invalid word is still considered to be invalid.

Gudahtt committed:
Merge branch 'develop' into extract-srp-input-from-create-vault
